### PR TITLE
Built and pushed Docker image to Github Packages

### PR DESCRIPTION
**Pull request by tobiasmcnulty**
This will be used to deploy the new Trac instance on 20.04, and might be useful for local dev.

Note, an annoying side effect of this PR (and no env var support in Trac) is that local dev via Docker will end up with a spurious edit to `trac.ini`, due to needing to discover and configure the `database` from an environment variable in `docker-entrypoint.sh`:

```diff
$ git diff
diff --git a/trac-env/conf/trac.ini b/trac-env/conf/trac.ini
index ed4d0a3..4f57398 100644
--- a/trac-env/conf/trac.ini
+++ b/trac-env/conf/trac.ini
@@ -279,7 +279,7 @@ ticket_show_details = enabled
 auth_cookie_lifetime = 1209600
 base_url = https://code.djangoproject.com
 check_auth_ip = enabled
-database = postgres://code.djangoproject@/code.djangoproject
+database = postgres://code.djangoproject:code.djangoproject@db/code.djangoproject
 default_charset = utf-8
 default_handler = CustomWikiModule
 ignore_auth_case = disabled
```

I'm leaving the `database` value the way it is for now to avoid breaking existing local setups, but if everyone ends up using Docker for local dev anyways, we could change the default in `trac.ini` to the same value set in `docker-compose.yml`.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# The official python:2.7 image no longer receives automatic rebuilds (it's
+# a year old as of October, 2021), so use the latest LTS release of Ubuntu
+# that includes Python 2.7 instead.
+FROM ubuntu:20.04
+
+# Install packages needed to run your application (not build deps).
+RUN set -ex \
```
Comment (MarkusH) [inline Dockerfile:7]: Does one need the `set -ex`? Never seen it in `RUN` statements before

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    && RUN_DEPS=" \
+    ca-certificates \
+    git \
+    libpq5 \
+    make \
+    postgresql-client \
+    python2.7 \
+    " \
+    && apt-get update && apt-get install -y --no-install-recommends $RUN_DEPS \
```
Comment (MarkusH) [inline Dockerfile:16]: I'd avoid the variable definition and inline the dependencies. But that's a personal preference.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    build-essential \
+    libpq-dev \
+    python2.7-dev \
+    wget \
+    " \
+    && apt-get update && apt-get install -y --no-install-recommends $BUILD_DEPS \
+    && wget -q -O /tmp/get-pip.py https://bootstrap.pypa.io/pip/2.7/get-pip.py \
+    && python2.7 /tmp/get-pip.py \
+    && rm /tmp/get-pip.py \
+    && pip install virtualenv \
```
Comment (MarkusH) [inline Dockerfile:38]: Can we do `python2.7 -m pip install ...` instead. And similarly below `/venv/bin/python -m pip install ...`

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      dockerfile: Dockerfile
+    command: ["/venv/bin/tracd", "--port", "9000", "-s", "trac-env"]
+    volumes:
+      - ./trac-env/:/code/trac-env/
+    ports:
+      - 9000:9000
+    depends_on:
+      - db
+  db:
+    image: postgres:11.2-alpine
```
Comment (MarkusH) [inline docker-compose.yml:18]: Aren't we using PG 10 right now?

File: Dockerfile (hunk 3 above)
Comment (tobiasmcnulty) [reply Dockerfile:38]: I don't believe so, unfortunately...

```
$ podman run -it ubuntu:20.04
root@36152805e9bb:/# apt update
<snip>
root@36152805e9bb:/# apt install -y python2.7
<snip>
root@36152805e9bb:/# python2.7 -m pip
/usr/bin/python2.7: No module named pip
```

File: Dockerfile (hunk 1 above)
Comment (tobiasmcnulty) [reply Dockerfile]: I guess it was helpful to me for debugging at some point (and might be in the future), though the `-e` is indeed [redundant](https://stackoverflow.com/a/47143497/166053); I'll take that out.

File: docker-compose.yml (hunk 4 above)
Comment (tobiasmcnulty) [reply docker-compose.yml]: Ah, yes. Good catch. This was a copy-paste from dp.com, so that is likely incorrect too.

File: Dockerfile (hunk 3 above)
Comment (tobiasmcnulty) [reply Dockerfile:38]: Ah! It was just this line you're wondering about isn't it, not how it's installed. 😊 Yes we can definitely do that.

**Review by apollo13 — APPROVED** [review]
LGTM. The Dockerfile itself does imo not really use best practices but we probably don't really care because we do not develop on trac that much. If we were we might wanna make the images more cache efficient; but this is more than good enough for now I think.

As an example on how I usually structure docker files I can offer: https://gitlab.com/rocketduck/csi-plugin-nfs/-/blob/main/Dockerfile

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    image: ghcr.io/django/code.djangoproject.com:latest
+    # Comment out to use the ghcr.io image:
+    build:
+      context: ./
+      dockerfile: Dockerfile
+    command: ["/venv/bin/tracd", "--port", "9000", "-s", "trac-env"]
+    environment:
+     - TRAC_INI_database=postgres://code.djangoproject:code.djangoproject@db/code.djangoproject
+    volumes:
+      - ./trac-env/:/code/trac-env/
```
Comment (apollo13) [inline docker-compose.yml:14]: We might wanna provide a way to map the source in for easier development

**Review by felixxm — COMMENTED** [review]
@tobiasmcnulty Many thanks for all your efforts :1st_place_medal: 

I couldn't force `podman` to rebuild my images so I ended with `podman image prune --all` :rofl: Is there a way to force `podman` to rebuild images on changes? :thinking: 

Can we add `trac-admin "/code/trac-env" upgrade` to the `Dockerfile`? I'm getting the following error on `localhost:9000`:
```
Trac Error

TracError: The Trac Environment needs to be upgraded. Run:

  trac-admin "/code/trac-env" upgrade
```

File: docker-compose.yml (hunk 5 above)
Comment (tobiasmcnulty) [reply docker-compose.yml:14]: Excellent point, I'll make this change.

**Comment by tobiasmcnulty** [comment]
> LGTM. The Dockerfile itself does imo not really use best practices but we probably don't really care because we do not develop on trac that much. If we were we might wanna make the images more cache efficient; but this is more than good enough for now I think.
> 
> As an example on how I usually structure docker files I can offer: https://gitlab.com/rocketduck/csi-plugin-nfs/-/blob/main/Dockerfile

Thanks for the review + example! The base image updates will probably bust the cache more often than anything else, and I like the smaller image size without the build deps installed. A cleaner way to solve the build deps problem is certainly with a multi-stage build, but yes perhaps we could do that in a future iteration...

**Comment by tobiasmcnulty** [comment]
> @tobiasmcnulty Many thanks for all your efforts 🥇
> 
> I couldn't force `podman` to rebuild my images so I ended with `podman image prune --all` 🤣 Is there a way to force `podman` to rebuild images on changes? 🤔
> 
> Can we add `trac-admin "/code/trac-env" upgrade` to the `Dockerfile`? I'm getting the following error on `localhost:9000`:
> 
> ```
> Trac Error
> 
> TracError: The Trac Environment needs to be upgraded. Run:
> 
>   trac-admin "/code/trac-env" upgrade
> ```

Thanks for testing this! As mentioned in IRC, I think I found the underlying cause of #93 and propose [a fix](https://github.com/django/djangoproject.com/pull/1147) on the dp.com repo. I also added instructions to this README for updating that file in the future.

**Pull request merged**